### DISCUSSION TITLE
branch based auto-release

### DISCUSSION THREAD

**Discussion by mrmiroslav** [semantic-release/semantic-release#1786] [Q&A] [answered]
From what I have seen in the docs so far, MAJOR / PATCH / MINOR version is based on Angular Commit Message Conventions.

Can it be branch based, rather than commit message. For example:

patch/bug-fix (could allow fix/bug-fix too)
minor/new-feature (or feature/new-feature)
major/breaking-change 

Commit messages reflects changes, and not releases. 
It is easier to see upcoming releases by just looking at branches.

**Comment by travi** [comment, marked as answer]
it sounds like you may be either misunderstanding the goals of semantic-release or looking for a different tool. here are a few details that may help to clarify.

first, semantic-release strongly encourages a continuous-delivery/continuous-deployment style of workflow, with the goal being that fixes/features that are complete should be made available to consumers as soon as possible. as part of that process, the tool makes it as simple as possible to communicate the impact of the changes to the consumer (breaking/new feature/fix). it accomplishes the latter goal by asking the package maintainer to describe the impact of each commit, so that the appropriate version bump can be applied, based on the list of unreleased changes, at the time of release (usually when the commits make it to the mainline branch).

until those commits make it to a release branch, like your mainline branch, no release occurs. it is up to you what branching strategy you want to use for getting your commits ready for commits. until they make it to a release branch, there is really no concept of a release, only changes, each with a certain impact level.

the difficulty with attempting to describe the impact of changes based on branches is that you often do not know the impact at the time of creating a branch. describing impact at commit time is more optimal because the change has already been made and it can then be described more effectively.

now, with all of that context, we do support a workflow that sounds somewhat along the lines of what you describe with [pre-releases](https://semantic-release.gitbook.io/semantic-release/usage/workflow-configuration#pre-release-branches), but the branches describe the stability of the change, not the impact. the impact is still expected to be described in each commit so that the overall impact can influence the version bump appropriately at release time.

> It is easier to see upcoming releases by just looking at branches.

i'm not quite sure what your goal is with "upcoming releases", but the summary of my suggestion would be this: 
* if changes are stable, semantic-release automates the release process well enough that you can release immediately and not have to worry about "upcoming" anymore. consumers can benefit from the changes earlier
* if there is instability in the changes and you need to prove some things out before releasing a stable version to consumers, pre-releases enable you to still publish the changes immediately, but not to the latest/stable channel, so you can install the less stable versions in whatever consumers you need to test to ensure stability without impacting all consumers

**Comment by dominykas** [comment]
While I agree with the post above, just want to note, that if you really have your mind set on it, you _could_ write your own analyzer plugin and make it bump the version based on whatever rules you define (e.g. we have plugins at work for bumping explicitly based on an environment variable [when you need to re-release some docker containers without adding new code] and for bumping based on github PR label [for folks who don't want to follow the commit convention]).

https://semantic-release.gitbook.io/semantic-release/usage/plugins
